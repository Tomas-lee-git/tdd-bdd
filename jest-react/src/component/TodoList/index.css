.list {
  display: flex;
  justify-content: center;
  align-items: center;
  margin-bottom: 20px;
}
.list-container {
  width: 500px;
}
.list-title {
  font-size: 20px;
  font-weight: bold;
  display: flex;
  justify-content: space-between;
}
.list-counter {
  font-size: 14px;
  color: #666;
  background-color: #e6e6fa;
  border-radius: 50%;
  line-height: 20px;
  height: 20px;
  padding: 0 5px;
  text-align: center;
}
.list-content {
  list-style-type: none;
}
.list-item {
  line-height: 24px;
  height: 24px;
  font-size: 24px;
  margin-bottom: 10px;
  background-color: #fff;
  border-left: 3px solid #629a9c;
  text-indent: 10px;
  display: flex;
  align-items: center;
  justify-content: space-between;
  border-radius: 3px;
}
.delete-item {
  margin-right: 10px;
  background: #e6e6ea;
  text-align: center;
  width: 20px;
  height: 20px;
  line-height: 20px;
  border-radius: 10px;
  text-indent: 0px;
}
.list-div-item {
  width: 400px;
}
.list-input-item {
  width: 400px;
  height: 70%;
  text-indent: 10px;
  font-size: 20px;
  outline: none;
  border: none;
}
