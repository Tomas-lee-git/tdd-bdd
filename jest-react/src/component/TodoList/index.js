import React, { Component } from 'react';
import './index.css';
class TodoList extends Component {
  handleInputItemChange = (e, index) => {
    const { value } = e.target;
    const { handleModifiedItem } = this.props;
    handleModifiedItem(index, value);
  };

  handleClickWithEnter = e => {
    const { keyCode } = e;
    const { handleBlur } = this.props;
    if (keyCode === 13) {
      handleBlur();
    }
  };

  render() {
    const {
      title,
      todoList,
      deleteItem,
      handleFocus,
      handleBlur,
      handleChecked,
    } = this.props;
    return (
      <div className="list">
        <div className="list-container">
          <div className="list-title">
            <span data-test="listTitle">{title}</span>
            <span className="list-counter" data-test="counter">
              {todoList.length}
            </span>
          </div>
          <ul className="list-content">
            {todoList.map((item, index) => (
              <li
                key={`${index}-${item}`}
                className="list-item"
                data-test="listItem"
                onClick={() => {
                  handleFocus(index);
                }}
                onBlur={() => {
                  handleBlur();
                }}
              >
                <input
                  type="checkbox"
                  className="list-checkbox-item"
                  data-test="checkboxItem"
                  onClick={() => {
                    handleChecked(index);
                  }}
                />
                {item.isFocus ? (
                  <input
                    autoFocus
                    data-test="inputItem"
                    value={item.value}
                    className="list-input-item"
                    onChange={e => this.handleInputItemChange(e, index)}
                    onKeyUp={e => {
                      this.handleClickWithEnter(e, index);
                    }}
                  ></input>
                ) : (
                  <div data-test="divItem" className="list-div-item">
                    {item.value}
                  </div>
                )}
                <div
                  className="delete-item"
                  data-test="deleteItem"
                  onClick={e => {
                    e && e.stopPropagation();
                    deleteItem(index);
                  }}
                >
                  -
                </div>
              </li>
            ))}
          </ul>
        </div>
      </div>
    );
  }
}
export default TodoList;
