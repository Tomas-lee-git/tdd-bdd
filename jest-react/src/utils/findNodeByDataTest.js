const findNodeByDataTest = (wrapper, tag) => wrapper.find(`[data-test='${tag}']`);
export { findNodeByDataTest };
