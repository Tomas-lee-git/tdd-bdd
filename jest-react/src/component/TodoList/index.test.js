import React from 'react';
import { shallow } from 'enzyme';

import TodoList from './index';
import { findNodeByDataTest } from '../../utils/findNodeByDataTest';

describe('TodoList component test', () => {
  describe('dom test', () => {
    test('counter text should be 0, and todoList length should be 0', () => {
      const wrapper = shallow(<TodoList todoList={[]} />);
      const counter = findNodeByDataTest(wrapper, 'counter');
      expect(counter.text()).toEqual('0');
      const undoItems = findNodeByDataTest(wrapper, 'undoItem');
      expect(undoItems.length).toEqual(0);
    });

    test('counter text should be 3, undo list length should be 3', () => {
      const mockTodoList = [
        { isFocus: false, value: 'hello' },
        { isFocus: false, value: 'hello' },
        { isFocus: false, value: 'hello' },
      ];
      const wrapper = shallow(<TodoList todoList={mockTodoList} />);
      const counter = findNodeByDataTest(wrapper, 'counter');
      expect(counter.text()).toEqual('3');
      const undoItems = findNodeByDataTest(wrapper, 'undoItem');
      expect(undoItems.length).toEqual(3);
    });

    test("should have one inputItem when item's isFocus is true", () => {
      const mockTodoList = [
        { isFocus: false, value: 'hello' },
        { isFocus: true, value: 'hello' },
        { isFocus: false, value: 'hello' },
      ];
      const wrapper = shallow(<TodoList todoList={mockTodoList} />);
      const inputItems = findNodeByDataTest(wrapper, 'inputItem');
      expect(inputItems.length).toEqual(1);
    });

    test('should have specific title', () => {
      const  title = 'special title';
      const wrapper = shallow(<TodoList  todoList={[]} title={title}/>);
      const listTitles = findNodeByDataTest(wrapper, 'listTitle');
      expect(listTitles.at(0).text()).toEqual(title);
    });

    test('snapshot test', () => {
      const wrapper = shallow(<TodoList todoList={[]} />);
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('function test', () => {
    describe('delete function test', () => {
      test('should have corresponding delete button when list exist', () => {
        const mockTodoList = [
          { isFocus: false, value: 'hello' },
          { isFocus: false, value: 'hello' },
          { isFocus: false, value: 'hello' },
        ];
        const wrapper = shallow(<TodoList todoList={mockTodoList} />);
        const deleteItems = findNodeByDataTest(wrapper, 'deleteItem');
        expect(deleteItems.length).toEqual(3);
      });

      test('delete function should be called after click delete btn', () => {
        const mockTodoList = [
          { isFocus: false, value: 'hello' },
          { isFocus: false, value: 'hello' },
          { isFocus: false, value: 'hello' },
        ];
        const mockDeleteItemFn = jest.fn();
        const wrapper = shallow(
          <TodoList todoList={mockTodoList} deleteItem={mockDeleteItemFn} />
        );
        const deleteItems = findNodeByDataTest(wrapper, 'deleteItem');
        const index = 1;
        deleteItems.at(index).simulate('click');
        expect(mockDeleteItemFn).toHaveBeenCalledWith(index);
        expect(mockDeleteItemFn).toHaveBeenCalledTimes(1);
      });
    });

    describe('click listItem test', () => {
      test('handleFocus fn should be called after click listItem', () => {
        const mockTodoList = [
          { isFocus: false, value: 'hello' },
          { isFocus: false, value: 'hello' },
          { isFocus: false, value: 'hello' },
        ];
        const mockHandleFocus = jest.fn();
        const wrapper = shallow(
          <TodoList todoList={mockTodoList} handleFocus={mockHandleFocus} />
        );
        const undoItems = findNodeByDataTest(wrapper, 'undoItem');
        const index = 1;
        undoItems.at(index).simulate('click');
        expect(mockHandleFocus).toHaveBeenCalledWith(index);
        expect(mockHandleFocus).toHaveBeenCalledTimes(1);
      });
    });

    describe('listItem blur test', () => {
      test('handleBlur should be called after listItem blur', () => {
        const mockTodoList = [
          { isFocus: false, value: 'hello' },
          { isFocus: false, value: 'hello' },
          { isFocus: false, value: 'hello' },
        ];
        const mockHandleBlur = jest.fn();
        const wrapper = shallow(
          <TodoList todoList={mockTodoList} handleBlur={mockHandleBlur} />
        );
        const undoItems = findNodeByDataTest(wrapper, 'undoItem');
        undoItems.at(0).simulate('blur');
        expect(mockHandleBlur).toHaveBeenCalledTimes(1);
      });
    });

    describe('inputItem change test', () => {
      test('modifyItem method should be called with value', () => {
        const mockTodoList = [
          { isFocus: true, value: 'hello' },
          { isFocus: false, value: 'hey' },
        ];
        const mockHandleModifyItem = jest.fn();
        const wrapper = shallow(
          <TodoList
            todoList={mockTodoList}
            handleModifiedItem={mockHandleModifyItem}
          />
        );
        const inputItems = findNodeByDataTest(wrapper, 'inputItem');
        const inputValue = 'hello,world';
        inputItems.at(0).simulate('change', {
          target: {
            value: inputValue,
          },
        });
        expect(mockHandleModifyItem).toHaveBeenCalledTimes(1);
        expect(mockHandleModifyItem).toHaveBeenCalledWith(0, inputValue);
      });
    });

    describe('inputItem press "enter" key test', () => {
      test('after inputItem press "Enter" key , handleBur function should be called', () => {
        const inputValue = 'hello, world';
        const mockTodoList = [{ isFocus: true, value: inputValue }];
        const mockHandleBlur = jest.fn();
        const wrapper = shallow(
          <TodoList todoList={mockTodoList} handleBlur={mockHandleBlur} />
        );
        const inputItems = findNodeByDataTest(wrapper, 'inputItem');
        inputItems.at(0).simulate('keyUp', {
          keyCode: 13,
        });
        expect(mockHandleBlur).toHaveBeenCalledTimes(1);
      });
    });
  });
});
