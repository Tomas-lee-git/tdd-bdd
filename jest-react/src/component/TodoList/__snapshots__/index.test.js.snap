// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList component test dom test snapshot test 1`] = `
<div
  className="list"
>
  <div
    className="list-container"
  >
    <div
      className="list-title"
    >
      <span
        data-test="listTitle"
      />
      <span
        className="list-counter"
        data-test="counter"
      >
        3
      </span>
    </div>
    <ul
      className="list-content"
    >
      <li
        className="list-item"
        data-test="listItem"
        key="0-[object Object]"
        onBlur={[Function]}
        onClick={[Function]}
      >
        <input
          className="list-checkbox-item"
          data-test="checkboxItem"
          type="checkbox"
        />
        <div
          className="list-div-item"
          data-test="divItem"
        >
          hello
        </div>
        <div
          className="delete-item"
          data-test="deleteItem"
          onClick={[Function]}
        >
          -
        </div>
      </li>
      <li
        className="list-item"
        data-test="listItem"
        key="1-[object Object]"
        onBlur={[Function]}
        onClick={[Function]}
      >
        <input
          className="list-checkbox-item"
          data-test="checkboxItem"
          type="checkbox"
        />
        <div
          className="list-div-item"
          data-test="divItem"
        >
          hello
        </div>
        <div
          className="delete-item"
          data-test="deleteItem"
          onClick={[Function]}
        >
          -
        </div>
      </li>
      <li
        className="list-item"
        data-test="listItem"
        key="2-[object Object]"
        onBlur={[Function]}
        onClick={[Function]}
      >
        <input
          className="list-checkbox-item"
          data-test="checkboxItem"
          type="checkbox"
        />
        <div
          className="list-div-item"
          data-test="divItem"
        >
          hello
        </div>
        <div
          className="delete-item"
          data-test="deleteItem"
          onClick={[Function]}
        >
          -
        </div>
      </li>
    </ul>
  </div>
</div>
`;
